x-pack/platform/plugins/shared/streams_app/public/components/stream_detail_significant_events_view/add_significant_event_flyout/generated_flow_form/significant_events_generated_table.tsx

```diff
@@ -15,45 +15,94 @@ import {
   type EuiTableSelectionType,
 } from '@elastic/eui';
 import { i18n } from '@kbn/i18n';
-import type { StreamQueryKql, Streams } from '@kbn/streams-schema';
-import React, { useState, type ReactNode } from 'react';
+import type { StreamQueryKql, Streams, System } from '@kbn/streams-schema';
+import React, { useCallback, useEffect, useState, type ReactNode } from 'react';
+import type { DataView } from '@kbn/data-views-plugin/public';
 import { PreviewDataSparkPlot } from '../common/preview_data_spark_plot';
 import { validateQuery } from '../common/validate_query';
 import { GeneratedEventPreview } from './generated_event_preview';
 
 interface Props {
   definition: Streams.all.Definition;
   generatedQueries: StreamQueryKql[];
+  setIsEditingQueries: (isEditingQueries: boolean) => void;
+  onEditQuery: (query: StreamQueryKql) => void;
   selectedQueries: StreamQueryKql[];
   isSubmitting: boolean;
   onSelectionChange: (selectedItems: StreamQueryKql[]) => void;
+  systems: Omit<System, 'description'>[];
+  dataViews: DataView[];
 }
 
 export function SignificantEventsGeneratedTable({
   generatedQueries,
+  onEditQuery,
+  setIsEditingQueries,
   selectedQueries,
   onSelectionChange,
   definition,
   isSubmitting,
+  systems,
+  dataViews,
 }: Props) {
   const [itemIdToExpandedRowMap, setItemIdToExpandedRowMap] = useState<Record<string, ReactNode>>(
     {}
   );
+  const [eventsInEditMode, setEventsInEditMode] = useState<string[]>([]);
+
+  const setIsEditing = useCallback(
+    (isEditing: boolean, query: StreamQueryKql) => {
+      const nextEventsInEditMode = isEditing
+        ? [...eventsInEditMode, query.id]
+        : eventsInEditMode.filter((id) => id !== query.id);
+      setEventsInEditMode(nextEventsInEditMode);
+      setIsEditingQueries(nextEventsInEditMode.length > 0);
+    },
+    [eventsInEditMode, setIsEditingQueries]
+  );
 
   const toggleDetails = (query: StreamQueryKql) => {
     const itemIdToExpandedRowMapValues = { ...itemIdToExpandedRowMap };
-    const validation = validateQuery(query);
 
     if (itemIdToExpandedRowMapValues[query.id]) {
       delete itemIdToExpandedRowMapValues[query.id];
     } else {
       itemIdToExpandedRowMapValues[query.id] = (
-        <GeneratedEventPreview definition={definition} query={query} validation={validation} />
+        <GeneratedEventPreview
+          definition={definition}
+          query={query}
+          isEditing={eventsInEditMode.includes(query.id)}
+          setIsEditing={(nextIsEditing) => setIsEditing(nextIsEditing, query)}
+          onSave={onEditQuery}
+          systems={systems}
+          dataViews={dataViews}
+        />
       );
     }
+
     setItemIdToExpandedRowMap(itemIdToExpandedRowMapValues);
   };
 
+  useEffect(() => {
```

Comment:
This is needed because the expanded rows of the basic table are not part of the normal render tree.
So a change in this parent component doesn't trigger the expanded row component to re-render (it only causes the table row itself to re-render), probably because the component is stored as a reference in the map `itemIdToExpandedRowMap`.

To work around this, when it needs to re-render, we assign a new value for each item in that map which causes the expanded row to re-render but no internal state is lost.